### PR TITLE
Binary/octal/hexadecimal string support

### LINKED ISSUE #345: hex representation
It would be nice to be able to convert mpmath floats to and from a string hex representation (like `float.hex` and `float.fromhex`).

### PR DESCRIPTION
**Pull request by warnerjon12**
Response to issue #345.

Closes #345

### CODE REVIEW

**Comment by warnerjon12** [comment]
I still need to add some tests for the new stuff and debug. This is not quite ready yet, but if you have any comments, feel free to state them

**Comment by fredrik-johansson** [comment]
I have not reviewed everything yet, but a first observation is that all the convenience methods are unnecessary. This is not user-facing code. The extra base keyword argument to `to_str` etc. should be sufficient.

A small optimization: in the line `while i >= 0 and digits[i] == stddigits[base-1]:` pull out `stddigits[base-1]` so that this doesn't have to be executed each iteration.

**Comment by warnerjon12** [comment]
I removed some of the convenience methods and did a few other things. Still need to add more tests

**Comment by warnerjon12** [comment]
At least all of the regular tests pass now

**Comment by skirpichev** [comment]
@warnerjon12, could you merge the master branch to trigger CI build?

File: mpmath/libmp/__init__.py [hunk 1]
```diff
@@ -11,8 +11,8 @@
   mpf_mul, mpf_mul_int, mpf_shift, mpf_frexp,
   mpf_div, mpf_rdiv_int, mpf_mod, mpf_pow_int,
   mpf_perturb,
-  to_digits_exp, to_str, str_to_man_exp, from_str, from_bstr, to_bstr,
-  mpf_sqrt, mpf_hypot)
+  to_digits_exp, to_str, to_bin, to_oct, to_hex, str_to_man_exp,
+  from_str, from_bin, from_oct, from_hex, mpf_sqrt, mpf_hypot)
```
Comment (skirpichev) [inline mpmath/libmp/__init__.py:15]: As it was noted by Fredrik, there is no need for additional methods

File: mpmath/libmp/libmpf.py [hunk 2]
```diff
@@ -1194,73 +1203,82 @@ def to_digits_exp(s, dps):
...
+    sb = bin_to_radix(sf, fixprec, base, fixps)
+    digits = numeral(sb, base=base, size=ps)
 
-    exponent += len(digits) - fixdps - 1
+    exponent += len(digits) - fixps - 1
     return sign, digits, exponent
 
-def to_str(s, dps, strip_zeros=True, min_fixed=None, max_fixed=None,
-    show_zero_exponent=False):
+def to_str(s, ps, strip_zeros=True, min_fixed=None, max_fixed=None,
```
Comment (skirpichev) [inline mpmath/libmp/libmpf.py:1214]: Why?  In fact, this is an API change: to_str() seems to be a public function.

**Comment by skirpichev** [comment]
There are test failures with gmpy2.

File: mpmath/libmp/libmpf.py [hunk 3]
```diff
@@ -327,7 +327,7 @@ def from_man_exp(man, exp, prec=None, rnd=round_fast):
 if BACKEND == 'sage':
     from_man_exp = sage_utils.from_man_exp
 
-def from_int(n, prec=0, rnd=round_fast):
+def from_int(n, prec=None, rnd=round_fast):
```
Comment (skirpichev) [inline mpmath/libmp/libmpf.py:330]: This breaks gmpy builds.
```suggestion
def from_int(n, prec=0, rnd=round_fast):
```

File: mpmath/libmp/libmpf.py [hunk 4]
```diff
@@ -1323,41 +1375,63 @@ def from_str(x, prec, rnd=round_fast):
...
+    specified direction if *prec* is not *None*. The literal syntax is
+    the same as for Python floats.
+    """
+    man, exp = str_to_man_exp(x, base=8)
+    return from_man_exp(man, (exp<<1)+exp, prec, rnd)
 
-def to_bstr(x):
-    sign, man, exp, bc = x
-    return ['','-'][sign] + numeral(man, size=bitcount(man), base=2) + ("e%i" % exp)
+def from_bin(x, prec=None, rnd=round_fast):
```
Comment (skirpichev) [inline mpmath/libmp/libmpf.py:1428]: This also breaks gmpy builds.
```suggestion
def from_bin(x, prec=0, rnd=round_fast):
```

**Comment by skirpichev** [comment]
Ok, #703 will include `from_str()`-stuff from this pr.

Lets close this stale pr.  Feel free to address pr review comments and re-open the pr.

**Pull request closed without merging**